File: articles/enzymes-regulation-biochemical-pathways/fortran/enzyme_kernel.f90
! Compact enzyme kinetic numerical kernel in Fortran.

program enzyme_kernel
  implicit none

  real :: S
  real :: Vmax
  real :: Km
  real :: I
  real :: Ki
  real :: P
  real :: Kf
  real :: v_control
  real :: v_competitive
  real :: v_noncompetitive
  real :: v_feedback
  real :: kcat
  real :: efficiency

  S = 10.0
  Vmax = 120.0
  Km = 5.0
  I = 4.0
  Ki = 2.0
  P = 8.0
  Kf = 6.0
  kcat = 75.0

  v_control = (Vmax * S) / (Km + S)
  v_competitive = (Vmax * S) / (Km * (1.0 + I / Ki) + S)
  v_noncompetitive = (Vmax / (1.0 + I / Ki)) * S / (Km + S)
  v_feedback = v_control / (1.0 + P / Kf)
  efficiency = kcat / Km

  print *, "Control velocity:", v_control
  print *, "Competitive inhibition velocity:", v_competitive
  print *, "Noncompetitive inhibition velocity:", v_noncompetitive
  print *, "Feedback velocity:", v_feedback
  print *, "Catalytic efficiency:", efficiency
end program enzyme_kernel

File: articles/enzymes-regulation-biochemical-pathways/c/enzyme_kernel.c
/*
 * Compact enzyme kinetic kernel in C.
 */

#include <math.h>
#include <stdio.h>

double michaelis_menten(double substrate, double vmax, double km) {
    return (vmax * substrate) / (km + substrate);
}

double competitive_inhibition(double substrate, double vmax, double km, double inhibitor, double ki) {
    return (vmax * substrate) / (km * (1.0 + inhibitor / ki) + substrate);
}

double noncompetitive_inhibition(double substrate, double vmax, double km, double inhibitor, double ki) {
    return (vmax / (1.0 + inhibitor / ki)) * substrate / (km + substrate);
}

double hill_response(double substrate, double vmax, double k, double n) {
    return vmax * pow(substrate, n) / (pow(k, n) + pow(substrate, n));
}

double catalytic_efficiency(double kcat, double km) {
    return kcat / km;
}

double feedback_velocity(double substrate, double product, double vmax, double km, double kf) {
    double base = michaelis_menten(substrate, vmax, km);
    return base / (1.0 + product / kf);
}

int main(void) {
    double substrate = 10.0;
    double vmax = 120.0;
    double km = 5.0;
    double inhibitor = 4.0;
    double ki = 2.0;
    double product = 8.0;
    double kf = 6.0;

    printf("michaelis_menten=%.6f\n", michaelis_menten(substrate, vmax, km));
    printf("competitive_inhibition=%.6f\n", competitive_inhibition(substrate, vmax, km, inhibitor, ki));
    printf("noncompetitive_inhibition=%.6f\n", noncompetitive_inhibition(substrate, vmax, km, inhibitor, ki));
    printf("hill_response=%.6f\n", hill_response(substrate, vmax, 6.0, 2.5));
    printf("catalytic_efficiency=%.6f\n", catalytic_efficiency(75.0, km));
    printf("feedback_velocity=%.6f\n", feedback_velocity(substrate, product, vmax, km, kf));

    return 0;
}
